osh/glob_.py

```diff
@@ -455,33 +501,37 @@ def _Glob(self, arg, out):
 
         return 0
 
-    def Expand(self, arg, out):
-        # type: (str, List[str]) -> int
+    def Expand(self, arg, out, blame_loc):
+        # type: (str, List[str], loc_t) -> int
         """Given a string that could be a glob, append a list of strings to
-        'out'.
+        'out' when glob expansion happens.
 
         Returns:
-          Number of items appended, or -1 for fatal failglob error.
+          Number of items appended, or -1 when glob expansion did not happen.
+        Raises:
+          error.FailGlob for fatal failglob error.
         """
+
         if self.exec_opts.noglob():
-            # we didn't glob escape it in osh/word_eval.py
-            out.append(arg)
-            return 1
+            # We cannot reconstruct the original string from the glob pattern,
```

Comment:
OK this part is interesting, let me think about whether there can be a different data structure